### ISSUE TITLE
Issue with regular arrays

### ISSUE THREAD

**Issue by luxint** [diku-dk/futhark#454]
Hi,

I'm making my first futhark programs and I like the language a lot! 

Nice and clean and so far I've not encountered any real issues ,except for one ...;-)
The constraint that arrays should be regular.. I can more or less work around it but the code gets contrived and ugly so I was wondering if anything can be done about it within the language itself. For instance the ability to iterate over a collection of 2 dimensional arrays would solve my problem. I know I could place them in a tuple but then you have to know statically how many components you need (which I don't) and you can't (AFAIK) iterate/access them dynamically with a variable.

What I've made is a multi level fully connected neural network (used to be called  a perceptron ;-) using stochastic gradient descent to solve. At the moment it is hard coded to three levels: input, hidden and output. Hard coded because of the limitation with arrays...I could work around it with flattening/splitting/reshaping/do matrix calculations/flatten/splitting  etc but this seems to me a very generic problem with would be better solved within the language.

**Comment by Athas** [comment]
I agree, the restriction to regular arrays can be problematic.  Unfortunately, it is not at all straightforward to solve - part of why Futhark performs so well is precisely because it carefully constrains programs in ways that ease the generation of efficient code.  We have vague ideas of what language support for irregular arrays may look like (basically extending the size annotations to permit a size to be an array of integers, rather than just a single integer), but few ideas on how to efficiently compile them.  Irregular arrays at the top level are one issue, but nested irregular arrays are particularly insidious.

For now, the solution, as you have discovered, is to keep a single concatenated array, and split it apart as necessary.  You can hide the complexity with some wrapper types; maybe something like:

```
-- Represents a sequence of 'n' arrays containing elements of type 't'.
type sequence [n] 't = { data: []t, starts: [n]i32, lengths: [n]i32 }

let sequence_singleton 't (x: []t): sequence [1] t =
  { data = x, starts = [0], lengths = [length x] }

let sequence_length [n] 't (_: sequence [n] t): i32 = n

let sequence_concat 't (s1: sequence [] t) (s2: sequence [] t): sequence [] t =
  { data = concat s1.data s2.data,
    starts = concat s1.starts (map (+length s1.data) s2.starts),
    lengths = concat s1.lengths s2.lengths }

let sequence_index 't (i: i32) (s: sequence [] t): []t =
  s.data[s.starts[i] : s.starts[i] + s.lengths[i]]
```

We are currently working on support higher-order functions, which would allow you to define a few more useful functions (right now you'd have to use a higher-order module), and then it would be almost tolerable.

Proper support for irregular arrays (including irregular parallelism) is unfortunately still a way off.  Many a PhD student has fallen on that quest, so we're trying a few other things first.

**Comment by luxint** [comment]
Thanks for the example, I was thinking along similar lines, but then for collection of 2d arrays/matrices: a tuple with a data array and an array with tuples containing the shapes (row/columns) of the data. To extract/get -> calculate start and end, split and  reshape  I would probably also add the possibility for negative indexes to access last entries. 

 Support for higher order functions would be really nice, that was actually another thing I missed ;-)
